## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<b>Задания:</b>

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Описание данных

<b>Таблица `books`</b>

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

<b>Таблица `authors`</b>

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

<b>Таблица `publishers`</b>

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

<b>Таблица `ratings`</b>

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

<b>Таблица `reviews`</b>

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Загрузка и обзор данных

In [1]:
#Иморт библиотек
import pandas as pd
from sqlalchemy import create_engine 

In [2]:
try:
    # устанавливаем параметры
    db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

    connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
    # сохраняем коннектор
    engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 
    print('Данные загружены')
except:
    print('Данные не загружены')

Данные загружены


In [3]:
#Функцию для выполнения запроса
def select(query):
    return pd.io.sql.read_sql(query, con = engine)

In [4]:
#Загрузка данных из таблицы
books = '''
SELECT * 
FROM books
'''
#Первые пять строк датафрейма
select(books).head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
#информация о датафрейме
select(books).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [6]:
#Загрузка данных из таблицы
authors = '''
SELECT * 
FROM authors
'''
#Первые пять строк датафрейма
select(authors).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
#информация о датафрейме
select(authors).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [8]:
#Загрузка данных из таблицы
publishers = '''
SELECT * 
FROM authors
'''
#Первые пять строк датафрейма
select(publishers).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
#информация о датафрейме
select(publishers).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [10]:
#Загрузка данных из таблицы
ratings = '''
SELECT * 
FROM authors
'''
#Первые пять строк датафрейма
select(ratings).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [11]:
#информация о датафрейме
select(ratings).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [12]:
#Загрузка данных из таблицы
reviews = '''
SELECT * 
FROM authors
'''
#Первые пять строк датафрейма
select(reviews).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [13]:
#информация о датафрейме
select(reviews).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### Вывод

Данные достаточно чистые, пропуски отсутвуют.

## Решение задач

### Посчитать, сколько книг вышло после 1 января 2000 года

In [14]:
count_book = ''' 
SELECT COUNT(*)
FROM books
WHERE publication_date::date > '01-01-2000'
'''

select(count_book)

count
0    819

После 1 января 2000 года вышла 819 книг.

### Для каждой книги посчитать количество обзоров и среднюю оценку

In [15]:
ratings = '''
SELECT books.book_id, 
       title,  
       COUNT(DISTINCT review_id) as cnt,
       ROUND(AVG(rating), 1) as rating
FROM books 
LEFT JOIN ratings ON books.book_id = ratings.book_id
LEFT JOIN reviews ON books.book_id = reviews.book_id
GROUP BY books.book_id
ORDER BY rating
'''

select(ratings)

book_id                                              title  cnt  rating
0        303                               Harvesting the Heart    2     1.5
1        371                                              Junky    2     2.0
2        202                                      Drowning Ruth    3     2.0
3        316                  His Excellency: George Washington    2     2.0
4        794                                  The Mermaid Chair    3     2.3
..       ...                                                ...  ...     ...
995      444       Moneyball: The Art of Winning an Unfair Game    3     5.0
996      169  Crucial Conversations: Tools for Talking When ...    2     5.0
997      642                 The Big Bad Wolf  (Alex Cross  #9)    2     5.0
998      732  The Ghost Map: The Story of London's Most Terr...    2     5.0
999      182                                         Dead Souls    2     5.0

[1000 rows x 4 columns]

In [16]:
ratings = '''
SELECT books.book_id, 
       title,  
       COUNT(DISTINCT review_id) as cnt,
       ROUND(AVG(rating), 1) as rating
FROM books 
LEFT JOIN ratings ON books.book_id = ratings.book_id
LEFT JOIN reviews ON books.book_id = reviews.book_id
GROUP BY books.book_id
ORDER BY cnt
'''

select(ratings)

book_id                                              title  cnt  rating
0         83    Anne Rice's The Vampire Lestat: A Graphic Novel    0     3.7
1        672   The Cat in the Hat and Other Dr. Seuss Favorites    0     5.0
2        221                          Essential Tales and Poems    0     4.0
3        808                            The Natural Way to Draw    0     3.0
4        191  Disney's Beauty and the Beast (A Little Golden...    0     4.0
..       ...                                                ...  ...     ...
995      733                          The Giver (The Giver  #1)    6     3.8
996      854                                           The Road    6     3.8
997      695  The Curious Incident of the Dog in the Night-Time    6     4.1
998      696             The Da Vinci Code (Robert Langdon  #2)    6     3.8
999      948                            Twilight (Twilight  #1)    7     3.7

[1000 rows x 4 columns]

Максимальный рейтинг - 5, минимальный средний рейтинг - 1.5.<br>
Максимальное количество - 7, минимальное - 0.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [17]:
pub_books = '''
SELECT subquery_B.publisher,
       subquery_A.books_cnt
FROM (SELECT publisher_id AS publisher_id,
             COUNT(book_id) AS books_cnt
      FROM books
      WHERE num_pages > 50
      GROUP BY publisher_id
      ORDER BY books_cnt DESC
      LIMIT 1) AS subquery_A

LEFT JOIN (SELECT publisher AS publisher,
           publisher_id AS publisher_id
           FROM publishers) AS subquery_B ON subquery_A.publisher_id = subquery_B.publisher_id
'''

select(pub_books)

publisher  books_cnt
0  Penguin Books         42

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books (42 книги).

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [18]:
ratings_author = '''
WITH
i AS (SELECT authors.author,
            books.book_id,
            AVG(ratings.rating) AS avg_rating
      FROM authors
      FULL JOIN books ON books.author_id = authors.author_id
      FULL JOIN ratings ON ratings.book_id = books.book_id
      GROUP BY author,
              books.book_id
      HAVING COUNT (ratings.rating) >= 50)

SELECT i.author,
       ROUND(AVG(i.avg_rating),2) AS avg_rating
FROM  i
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
'''

select(ratings_author)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.28

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré	(рейтинг 4.28).

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [19]:
avg_reviews = '''
WITH
i AS (SELECT COUNT(reviews.review_id) AS reviews_count
             FROM reviews
             WHERE username IN (SELECT username
                                FROM ratings
                                GROUP BY username
                                HAVING COUNT(rating_id) > 50)
             GROUP BY username)
             
SELECT ROUND(AVG(i.reviews_count), 0) AS avg_reviews
       FROM i
'''

select(avg_reviews)

avg_reviews
0         24.0

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

## Выводы

После 1 января 2000 года вышло 819 книг.

Максимальный рейтинг - 5, минимальный средний рейтинг - 1.5.<br>
Максимальное количество - 7, минимальное - 0.

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books (42 книги).

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré (рейтинг 4.28).

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.